File: app/src/main/java/xyz/b515/schedule/ui/view/WeekCoursesFragment.java
package xyz.b515.schedule.ui.view;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.util.DisplayMetrics;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import java8.util.stream.StreamSupport;
import xyz.b515.schedule.Constant;
import xyz.b515.schedule.R;
import xyz.b515.schedule.db.CourseManager;
import xyz.b515.schedule.entity.Course;
import xyz.b515.schedule.entity.Spacetime;

/**
 * Created by Yun on 2017.4.24.
 */

public class WeekCoursesFragment extends Fragment {
    View view;
    @BindView(R.id.header_weeks) LinearLayout headerWeeks;
    @BindView(R.id.header_time) LinearLayout headerTime;
    @BindView(R.id.courses) RelativeLayout coursesLayout;
    CourseManager manager;
    SharedPreferences preferences;
    private int baseHeight;
    private int baseWidth;

    public void onCreate(Bundle bundle) {
        super.onCreate(bundle);
    }

    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.fragment_week, container, false);
        ButterKnife.bind(this, view);
        manager = new CourseManager(getContext());
        preferences = PreferenceManager.getDefaultSharedPreferences(getContext());

        showHeaders();
        showCourses();
        return view;
    }

    private void showHeaders() {
        final int column = 7;
        final int row = 12;

        DisplayMetrics metrics = getResources().getDisplayMetrics();
        baseHeight = (int) Math.round(metrics.heightPixels * (1.0 / 15.0));
        baseWidth = (int) Math.round(metrics.widthPixels * (2.0 / 15.0));

        for (int i = 0; i < column; i++) {
            TextView tv = new TextView(getContext());
            tv.setText(getResources().getStringArray(R.array.weekdays)[i]);
            tv.setGravity(Gravity.CENTER);
            tv.setWidth(baseWidth);
            headerWeeks.addView(tv);
        }
        for (int i = 1; i <= row; i++) {
            TextView tv = new TextView(getContext());
            tv.setText(String.valueOf(i));
            tv.setGravity(Gravity.CENTER);
            tv.setHeight(baseHeight);
            headerTime.addView(tv);
        }
    }

    private void showCourses() {
        int currentWeek = preferences.getInt(Constant.CURRENT_WEEK, -1);
        List<Course> list = manager.getAllCourse();
        if (list != null) {
            StreamSupport.stream(list)
                    .flatMap(course -> StreamSupport.stream(course.getSpacetimes()))
                    .filter(spacetime -> spacetime.getStartWeek() <= currentWeek && spacetime.getEndWeek() >= currentWeek)
                    .forEach(this::addCourseToView);
        }
    }

    private void addCourseToView(Spacetime spacetime) {
        Course course = spacetime.getCourse();

        LinearLayout v = (LinearLayout) LayoutInflater.from(getContext()).inflate(R.layout.item_course_view, coursesLayout, false);
        TextView textView = ButterKnife.findById(v, R.id.course_name);
        textView.setText(course.getName() + "\n" + spacetime.getLocation());
        v.setBackgroundColor(course.getColor());
        v.setX((spacetime.getWeekday() - 1) * baseWidth);
        v.setY((spacetime.getStartTime() - 1) * baseHeight);
        coursesLayout.addView(v);
        v.getLayoutParams().height = (spacetime.getEndTime() - spacetime.getStartTime() + 1) * baseHeight;
        v.getLayoutParams().width = baseWidth;
        v.setOnClickListener(view -> {
            Intent intent = new Intent(getContext(), CourseDetailActivity.class);
            intent.putExtra(Constant.TOOLBAR_TITLE, false);
            intent.putExtra(Constant.COURSE_ID, course.getId());
            getContext().startActivity(intent);
        });
    }

}

File: app/src/main/kotlin/xyz/b515/schedule/ui/view/WeekCoursesFragment.kt
package xyz.b515.schedule.ui.view

import android.app.Fragment
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.preference.PreferenceManager
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import kotlinx.android.synthetic.main.fragment_week.view.*
import xyz.b515.schedule.Constant
import xyz.b515.schedule.R
import xyz.b515.schedule.db.CourseManager
import xyz.b515.schedule.entity.Spacetime

/**
 * Created by Yun on 2017.4.24.
 */

class WeekCoursesFragment : Fragment() {
    lateinit var manager: CourseManager
    lateinit var preferences: SharedPreferences
    private var baseHeight: Int = 0
    private var baseWidth: Int = 0

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_week, container, false)
        manager = CourseManager(context!!)
        preferences = PreferenceManager.getDefaultSharedPreferences(context)

        showHeaders(view)
        showCourses(view)
        return view
    }

    private fun showHeaders(view: View) {
        val column = 7
        val row = 12

        val metrics = resources.displayMetrics
        baseHeight = Math.round(metrics.heightPixels * (1.0 / 15.0)).toInt()
        baseWidth = Math.round(metrics.widthPixels * (2.0 / 15.0)).toInt()

        for (i in 0 until column) {
            val tv = TextView(context)
            tv.text = resources.getStringArray(R.array.weekdays)[i]
            tv.gravity = Gravity.CENTER
            tv.width = baseWidth
            view.header_weeks.addView(tv)
        }
        for (i in 1..row) {
            val tv = TextView(context)
            tv.text = i.toString()
            tv.gravity = Gravity.CENTER
            tv.height = baseHeight
            view.header_time.addView(tv)
        }
    }

    private fun showCourses(view: View) {
        val currentWeek = preferences.getInt(Constant.CURRENT_WEEK, -1) + 1
        val list = manager.getAllCourse()
        list.flatMap { it.spacetimes!! }
                .filter { currentWeek in it.startWeek..it.endWeek }
                .forEach { this.addCourseToView(it, view) }
    }

    private fun addCourseToView(spacetime: Spacetime, view: View) {
        val course = spacetime.course

        val v = LayoutInflater.from(context).inflate(R.layout.item_course_view, view.courses_layout, false) as LinearLayout
        val textView = v.findViewById<TextView>(R.id.course_name)
        textView.text = "${course.name}\n${spacetime.location}"
        v.setBackgroundColor(course.color)
        val weekday = if (spacetime.weekday == 1) 7 else spacetime.weekday - 1
        v.x = ((weekday - 1) * baseWidth).toFloat()
        v.y = ((spacetime.startTime - 1) * baseHeight).toFloat()
        view.courses_layout.addView(v)
        v.layoutParams.height = (spacetime.endTime - spacetime.startTime + 1) * baseHeight
        v.layoutParams.width = baseWidth
        v.setOnClickListener {
            val intent = Intent(context, CourseDetailActivity::class.java)
            intent.putExtra(Constant.TOOLBAR_TITLE, false)
            intent.putExtra(Constant.COURSE_ID, course.id)
            context!!.startActivity(intent)
        }
    }

}
